### PR TITLE
Wrap LLVMAttributeRef API

### PR DESCRIPTION
**Pull request by kyouko-taiga**
**Disclaimer:** I'm quite a newbie with LLVM

This is my attempt at wrapping the LLVM 4.0.0+ `LLVMAttributeRef` API (as mentioned in #10).

I extracted the list of parameters from [LLVM's docs](http://llvm.org/docs/LangRef.html#function-attributes). I didn't make any unit tests so far, and haven't tested all parameters.

Due to my limited knowledge of LLVM, I'm unsure of the expected result of calling `LLVMAddAttributeAtIndex` on parameters whose arguments aren't a single integer value (e.g. `allocsize`), nor of the difference between attributes specified with or without quote in the documentation (e.g. `"patchable-function"`). Some review from a more experienced developer would be greatly appreciated.

I tried to follow the coding convention I could guess from the existing code to the best of my ability, but don't hesitate to comment on that as well.

### CODE REVIEW

File: Sources/LLVM/Function+Attributes.swift [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+extension Function {
+  /// Adds an attribute to the function.
+  ///
+  /// - parameter attr: The attribute to add.
+  /// - parameter value: The optional value of the attribute.
+  public func addAttribute(_ attr: FunctionAttribute, value: UInt64 = 0) {
+    let ctx = LLVMGetModuleContext(LLVMGetGlobalParent(llvm))
+    let attrRef = LLVMCreateEnumAttribute(ctx, attr.kindID, value)
+    LLVMAddAttributeAtIndex(llvm, 0, attrRef)
```
Comment (CodaFi) [inline Sources/LLVM/Function+Attributes.swift]: This is not correct.  This is the index for the return value of the function.

File: Sources/LLVM/Function+Attributes.swift [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+  }
+
+  /// Adds an attribute to the given function's parameter.
+  ///
+  /// - parameter attr: The attribute to add.
+  /// - parameter value: The optional value of the attribute.
+  /// - parameterIndex: The index of the parameter to which add the
+  ///   attribute, starting from 0.
+  public func addParameterAttribute(_ attr: ParameterAttribute, value: UInt64 = 0,
+                                    for parameterIndex: LLVMAttributeIndex) {
```
Comment (CodaFi) [inline Sources/LLVM/Function+Attributes.swift]: We have a policy of never exposing LLVM's types to the user.  I have an idea of how we can lift this into an enum in one case.

**Comment by CodaFi** [comment]
I think we can kill two birds with one stone here by lifting `AttributeList`'s sorta ad-hoc indexing model to an enum:

```swift
enum AttributeIndex: UInt32 {
  case returnValue // 0
  case function // ~0
}
```

The `RawRepresentable` conformance will have to be done by hand.

File: Sources/LLVM/Function+Attributes.swift [hunk 3]
```diff
@@ -0,0 +1,262 @@
+#if SWIFT_PACKAGE
+import cllvm
+#endif
+
+/// Enumerates the attributes of LLVM functions.
+public enum FunctionAttribute: String {
```
Comment (CodaFi) [inline Sources/LLVM/Function+Attributes.swift]: Thank you for documenting this patch.

File: Sources/LLVM/Function+Attributes.swift (hunk 2 above)
Comment (CodaFi) [reply Sources/LLVM/Function+Attributes.swift]: Please use `Int` here.

**Comment by CodaFi** [comment]
Do you know if `lli` or the AttributeSet does bounds checking for the parameter indices?  If not we should put in a precondition.

File: Sources/LLVM/Function+Attributes.swift [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+  /// In some parallel execution models, there exist operations that cannot be
+  /// made control-dependent on any additional values.
+  case convergent
+  /// This attribute indicates that the function may only access memory that is
+  /// not accessible by the module being compiled.
+  case inaccessiblememonly
+  /// This attribute indicates that the function may only access memory that is
+  /// either not accessible by the module being compiled, or is pointed to by
+  /// its pointer arguments.
+  case inaccessiblemem_or_argmemonly
```
Comment (CodaFi) [inline Sources/LLVM/Function+Attributes.swift]: All the `snake_case` should be `camelCase`.

**Comment by CodaFi** [comment]
Oh, it'd also be great if you could add a test.  It can be as simple as a loop that adds/removes All The Attributes to a dummy function and its parameters and checks that they got registered/remoed.

**Comment by kyouko-taiga** [comment]
Thanks for the feedbacks!

A few comments on the changes I made:

According to [LLVM docs](https://llvm.org/docs/HowToUseAttributes.html), I understand there are several kinds of attributes in LLVM's new scheme. If I understand the docs correctly, this is also where the distinction between attributes specified with or without quotes comes in. The formers describe target-independent attributes, while the latters describe dependent ones. Therefore I took the liberty to create two types to distinguish between those two kind of attributes. I wrapped all `LLVM*StringAttribute*` functions, which I believe were part of LLVMSwift missing functionalities, as mentioned in #51 .

I implemented a method to retrieve the attributes from a function. I made the choice to return an array rather than a set, so as to avoid having to conform to `Hashable` (which seemed a bit overkill to me). Unfortunately, it seems like there's no wrapper around C++'s [`Attribute::getKindAsString()`](http://llvm.org/doxygen/classllvm_1_1Attribute.html#a3ad8f83f8d6165314fe8f173645dd458) so I can't provide a proper implementation of [`EnumAttribute.name`](https://github.com/kyouko-taiga/LLVMSwift/blob/880fbbe63ce66118b081196dcb8e62e89aae1a43/Sources/LLVM/Function%2BAttributes.swift#L243).

I wrote some tests for all these shenanigans. Those revealed that there seems to be a problem with the removal of valued enum attributes. I tried to understand what's causing this bug within LLVM's source, but I have to admit to find the whole thing quite puzzling. Indeed, removing an integer attribute seems to always fail, because [`AttributeList::removeAttribute()`](http://llvm.org/doxygen/classllvm_1_1AttributeSet.html#a1247d3316a7816eaee459556671caab5) attempts to add the attribute about to be removed without preserving its value:
```c++
AttributeSet AttributeSet::removeAttribute(LLVMContext &C,
                                             Attribute::AttrKind Kind) const {
  if (!hasAttribute(Kind)) return *this;
  AttrBuilder B;
  B.addAttribute(Kind);
  return removeAttributes(C, B);
}
```
This in turn fails in `AttrBuilder::addAttribute()` violating the second assertion:
```c++
AttrBuilder &AttrBuilder::addAttribute(Attribute::AttrKind Val) {
  assert((unsigned)Val < Attribute::EndAttrKinds && "Attribute out of range!");
  assert(Val != Attribute::Alignment && Val != Attribute::StackAlignment &&
         Val != Attribute::Dereferenceable && Val != Attribute::AllocSize &&
         "Adding integer attribute without adding a value!");
  Attrs[Val] = true;
  return *this;
}
```
Maybe an issue with my version of LLVM ...

**Comment by kyouko-taiga** [comment]
I confirm the bug about valued enum attributes removal is caused by LLVM: [[Attributes] Fix crash when attempting to remove alignment from an attribute list/set](https://github.com/llvm-mirror/llvm/commit/9bc0b1080f195636fed019bce979aa72892d6c69#diff-19a9dba7179bdf5da40ec95b8bfefe82).

File: Sources/LLVM/Function+Attributes.swift [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+  }
+
+  /// The kind ID of the attribute.
+  internal var kindID: UInt32 {
+    return LLVMGetEnumAttributeKind(llvm)
+  }
+
+  /// The name of the attribute's kind.
+  public var name: String {
+    return ""
```
Comment (CodaFi) [inline Sources/LLVM/Function+Attributes.swift]: If we can't provide this then we should omit it.

File: Sources/LLVM/Function+Attributes.swift [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+    switch self {
+    case .function: return ~0
+    case .returnValue: return 0
+    case .argument(let i): return UInt32(i) + 1
+    }
+  }
+}
+
+/// An LLVM attribute.
+public protocol Attribute {
```
Comment (CodaFi) [inline Sources/LLVM/Function+Attributes.swift]: Can you document this and its members a little more

File: Sources/LLVM/Function+Attributes.swift [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+  public func removeAttribute(_ attr: Attribute, from index: AttributeIndex) {
+    switch attr {
+    case let enumAttr as EnumAttribute:
+      LLVMRemoveEnumAttributeAtIndex(llvm, index.rawValue, enumAttr.kindID)
+    case let stringAttr as StringAttribute:
+      var length: UInt32 = 0
+      let cstring = LLVMGetStringAttributeKind(stringAttr.llvm, &length)
+      LLVMRemoveStringAttributeAtIndex(llvm, index.rawValue, cstring, length)
+    default:
+      fatalError()
```
Comment (CodaFi) [inline Sources/LLVM/Function+Attributes.swift]: Please add an explanation string to this fatal error

**Review by CodaFi — CHANGES_REQUESTED** [review]

**Comment by CodaFi** [comment]
Down to nits.  This is looking quite good.

**Comment by kyouko-taiga** [comment]
What would be the best thing to do about the issue with `removeAttribute(:from:)` on valued enum attributes? The issue seems to be [fixed in LLVM 7.0.0svn](http://llvm.org/doxygen/classllvm_1_1AttributeSet.html#a1247d3316a7816eaee459556671caab5) but if I remove the guard in [`IRAttributesSpec.swift:138`](https://github.com/kyouko-taiga/LLVMSwift/blob/71095452fdfa5358bf19b686d3a465b8760da2d0/Tests/LLVMTests/IRAttributesSpec.swift#L138) the tests won't pass with previous versions. I don't know what's LLVMSwift's stance about the version of LLVM it is linked to.

**Comment by CodaFi** [comment]
If we can't safely provide this API then we shouldn't provide it altogether.  We can hold out until 7.0 and leave behind a note.  I'd much rather that than crashing deep in LLVM.

**Comment by CodaFi** [comment]
What I would recommend is we bake the guard into the API and fatal error with an explanation as to why we cannot safely remove the attribute yet.

**Comment by kyouko-taiga** [comment]
Is there any way we can know what version of LLVM we're linked to?

**Comment by CodaFi** [comment]
`LLVM_VERSION_MAJOR/MINOR/PATCH` but we tie ourselves to the latest version as it is announced and do not do backwards compatibility.

**Comment by kyouko-taiga** [comment]
Alright, I'd make the guard only for the attributes that may fail, namely `align`, `alignstack`, `allocsize`, `dereferenceable` and `dereferenceable_or_null`.

**Review by CodaFi — APPROVED** [review]

**Comment by CodaFi** [comment]
Excellent.  Thank you kindly.

⛵️

**Pull request merged**